## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [356]:
import pandas as pd

df = pd.read_csv('client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [357]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [358]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [359]:
# Use this space to do any additional research
# and familiarize yourself with the data.

q1 = df["unit_cost"].quantile(0.25)
df.head()
print(q1)

14.84


In [360]:
#sum_of_empty = df.isnull().sum()
data_shape = df.shape

#sum_of_empty
data_shape

(54639, 18)

In [361]:
# What three item categories had the most entries?
df["category"].value_counts().head(3)



category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [362]:
# For the category with the most entries,
# which subcategory had the most entries?
df["subcategory"].value_counts().head()
#df.subcatogery.loc = [df.category == 'consumsbles'.vale_counts().,head(1)

df.subcategory[df.category == "consumables"].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [363]:
# Which five clients had the most entries in the data?
top_clients = df["client_id"].value_counts().head(5)
print(top_clients)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [364]:
# Store the client ids of those top 5 clients in a list.
top_clients = df["client_id"].value_counts().head(5).index.tolist()
print(top_clients)


[33615, 66037, 46820, 38378, 24741]


In [365]:
# How many total units (the qty column) did the
# client with the most entries order order?

# Get the client ID with the most entries
top_client_id = df["client_id"].value_counts().idxmax()

# Filter the DataFrame for the client with the most entries
top_client_data = df[df["client_id"] == top_client_id]

# Calculate the total units ordered by the top client
total_units_ordered = top_client_data["qty"].sum()

print(f"The client with the most entries ordered a total of {total_units_ordered} units.")


The client with the most entries ordered a total of 64313 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [366]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df['subtotal'] = df['unit_price'] * df['qty']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [367]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

ship_price = df.apply(lambda row: row['unit_weight'] * 7 
                    if row['unit_weight'] * row['qty'] > 50 
                    else row['unit_weight'] * 10, axis=1)

df['shipping_price'] = ship_price * df['qty']

df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20

In [368]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.0925
df['total_price_rounded'] = round(df['total_price'], 2)
df[['total_price','subtotal','shipping_price']].head()


total_price   subtotal  shipping_price
0  131839.076250  115164.00         5512.50
1     914.258625     523.95          312.90
2    1077.117600     527.28          458.64
3    1543.571400    1056.18          356.70
4   11364.403500    3902.00         6500.20

In [369]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df['cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df[['cost']].head()

cost
0  85597.05
1    629.79
2    765.18
3   1077.35
4   8663.60

In [370]:
# Create a column for the profit of
# each line using line cost and line price

df['profit'] = df['total_price'] - df['cost']
df[['profit']].head()

profit
0  46242.026250
1    284.468625
2    311.937600
3    466.221400
4   2700.803500

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [371]:
total_price_order_2742071 = df[df['order_id'] == 2742071]['total_price'].sum()
total_price_order_2173913 = df[df['order_id'] == 2173913]['total_price'].sum()
total_price_order_6128929 = df[df['order_id'] == 6128929]['total_price'].sum()

total_price_order1 = round(total_price_order_2742071, 2)
total_price_order2 = round(total_price_order_2173913, 2)
total_price_order3 = round(total_price_order_6128929, 2)

print(f'For Order ID 2742071, the total price is ${total_price_order1:.2f}')
print(f'For Order ID 2173913, the total price is ${total_price_order2:.2f}')
print(f'For Order ID 6128929, the total price is ${total_price_order3:.2f}')


For Order ID 2742071, the total price is $152811.90
For Order ID 2173913, the total price is $162388.72
For Order ID 6128929, the total price is $923441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [372]:
df['profit']


0        46242.026250
1          284.468625
2          311.937600
3          466.221400
4         2700.803500
             ...     
54634     1342.321200
54635     2717.320275
54636     8130.646250
54637     1352.713600
54638     3112.331250
Name: profit, Length: 54639, dtype: float64

In [373]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

#client_spent = df.groupby('client_id')['qty'].sum()
top_5_clients_by_quantity_spent = total_units_ordered.sum()
print(top_5_clients_by_quantity_spent)


64313


In [374]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.


summary_df = pd.DataFrame({
    'Total Units Purchased': df['qty'],
    'Total Shipping Price': df['shipping_price'],
    #'Total Revenue: df['...'], I got confused with the difference between total revenue and profit...looks like I did not calculate revenue 
    'Total Profit': df['profit']
})
                          
# Display the summary DataFrame
#print(summary_df)

summary_df.head()

Total Units Purchased  Total Shipping Price  Total Profit
0                    105               5512.50  46242.026250
1                     21                312.90    284.468625
2                     39                458.64    311.937600
3                     29                356.70    466.221400
4                     20               6500.20   2700.803500

In [375]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

df.rename(columns = {'first':'First','last':'Last', 'job':'Job', 'phone':'Phone', 'email':'Email', 'client_id':'Client ID', 'order_id':'Order ID', 'order_date':'Order Date', 'order_week':'Order Week', 'order_week':'Order Week', 'order_year':'Order Year', 'item_id':'Item ID', 'category':'Category', 'subcategory':'Subcategory', 'unit_price':'Unit Price', 'unit_cost':'Unit Cost', 'unit_weight':'Unit Weight', 'qty':'Quantity'})

First       Last                      Job               Phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     Email  Client ID  Order ID  Order Date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       Order Week  Order Year  ... Unit Cost Unit Weight Quantity  \
0              17        2023  ...    762.71        7.50      105   
1              20        2023  ...     15.09        1.49       21   
2               4        2023  ...      7.86        1.68       39   
3              17        2023  ...     24.85        1.23       29   
4              21        2023  ...    108.17       46.43       20   
...           ...         ...  ...       ...         ...      ...   
54634          17        2023  ...     51.60        2.25       33   
54635           6        2023  ...    175.46       11.70       47   
54636           9        2023  ...     57.31        4.16      475   
54637           3        2023  ...      1.22       18.04      112   
54638           1        2023  ...      1.61        2.07     1031   

       line_number   subtotal  shipping_price    total_price  \
0                1  115164.00         5512.50  131839.076250   
1                0     523.95          312.90     914.258625   
2                6     527.28          458.64    1077.117600   
3                3    1056.18          356.70    1543.571400   
4                1    3902.00         6500.20   11364.403500   
...            ...        ...             ...            ...   
54634            8    2743.29          519.75    3564.871200   
54635            0    9709.73         3849.30   14813.240275   
54636            6   31188.50        13832.00   49184.896250   
54637            0     165.76        14143.36   15632.713600   
54638            5    3103.31        14939.19   19711.431250   

       total_price_rounded      cost        profit  
0                131839.08  85597.05  46242.026250  
1                   914.26    629.79    284.468625  
2                  1077.12    765.18    311.937600  
3                  1543.57   1077.35    466.221400  
4                 11364.40   8663.60   2700.803500  
...                    ...       ...           ...  
54634              3564.87   2222.55   1342.321200  
54635             14813.24  12095.92   2717.320275  
54636             49184.90  41054.25   8130.646250  
54637             156

In [376]:
df['Profit_in_mil'] = df['profit']/1000000 
df['Profit_in_mil'] = df['Profit_in_mil'].apply(lambda x: f'${x: .2f}M')
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_weight   qty line_number   subtotal  \
0              17        2023  ...        7.50   105           1  115164.00   
1              20        2023  ...        1.49    21           0     523.95   
2               4        2023  ...        1.68    39           6     527.28   
3              17        2023  ...        1.23    29           3    1056.18   
4              21        2023  ...       46.43    20           1    3902.00   
...           ...         ...  ...         ...   ...         ...        ...   
54634          17        2023  ...        2.25    33           8    2743.29   
54635           6        2023  ...       11.70    47           0    9709.73   
54636           9        2023  ...        4.16   475           6   31188.50   
54637           3        2023  ...       18.04   112           0     165.76   
54638           1        2023  ...        2.07  1031           5    3103.31   

       shipping_price    total_price  total_price_rounded      cost  \
0             5512.50  131839.076250            131839.08  85597.05   
1              312.90     914.258625               914.26    629.79   
2              458.64    1077.117600              1077.12    765.18   
3              356.70    1543.571400              1543.57   1077.35   
4             6500.20   11364.403500             11364.40   8663.60   
...               ...            ...                  ...       ...   
54634          519.75    3564.871200              3564.87   2222.55   
54635         3849.30   14813.240275             14813.24  12095.92   
54636        13832.00   49184.896250             49184.90  41054.25   
54637        14143.36   15632.713600             15632.71  14280.00   
54638        14939.19   19711.431250             19711.43  16599.10   

             profit  Profit_in_mil  
0      46242.026250        $ 0.05M  
1        284.468625        $ 0.00M  
2        311.937600        $ 0.00M  
3        466.221400        $ 0.00M  
4       2700.803500        $ 0.00M  
...             ...            ...  
54634   1342.321200        $ 0.00M  
54635   2717.320275        $ 0.00M  
54636   8130.6

In [377]:
# Sort the updated data by "Total Profit" form highest to lowest

df.rename(columns={'first':'First', 'last':'Last', 'job':'Job', 'phone':'Phone', 'email':'Email', 'client_id':'Client ID', 'order_id':'Order ID', 'order_date':'Order Date', 'order_week':'Order Week', 'order_year':'Order Year', 'item_id':'Item ID', 'category':'Category', 'subcategory':'Subcategory', 'unit_price':'Unit Price', 'unit_cost':'Unit Cost', 'unit_weight':'Unit Weight', 'qty':'Quantity'}, inplace=True)

df['Top_5_sorted'] = df['Profit_in_mil']

Top_5_sorted = df.sort_values(by='Profit_in_mil', ascending=False).head(5)

Top_5_sorted.head(5)



First       Last                                       Job  \
40161  Charles  Mccormick  Conservation officer, historic buildings   
11444    James     Brooks                        Charity fundraiser   
29737  William     Savage                         Minerals surveyor   
23532    Susan     Glover               Industrial/product designer   
26518   Taylor     Barker                              Tree surgeon   

                       Phone                              Email  Client ID  \
40161  001-667-790-6949x2885  mccormick_sign@potter-sanford.com      46571   
11444   001-866-994-6097x017            brooks.j@morgan-fox.com      18789   
29737           534.689.6772          savagewilliam@johnson.com      30339   
23532   001-874-645-1061x557                  sglover@ellis.com      20849   
26518       970.537.6082x931          t.barker@jones-ibarra.net      38974   

       Order ID  Order Date  Order Week  Order Year  ... Quantity line_number  \
40161   1553623  2023-05-22          21        2023  ...   127184           6   
11444   5037924  2023-05-29          22        2023  ...   109328           0   
29737   9375270  2023-02-14           7        2023  ...   100479           7   
23532   7126774  2023-01-08           1        2023  ...    39337           2   
26518   3566343  2023-01-18           3        2023  ...    75729           5   

          subtotal  shipping_price   total_price  total_price_rounded  \
40161  12648448.80     20912865.12  3.666574e+07          36665735.46   
11444  44497589.28      1920892.96  5.071219e+07          50712191.85   
29737  23084045.46     19349241.03  4.635837e+07          46358365.49   
23532  15359525.02        33043.08  1.681638e+07          16816380.65   
26518  16814109.87       705036.99  1.913967e+07          19139667.94   

              cost        profit  Profit_in_mil  Top_5_sorted  
40161  28552808.00  8.112927e+06        $ 8.11M       $ 8.11M  
11444  43363857.92  7.348334e+06        $ 7.35M       $ 7.35M  
29737  39282265.05  7.076100e+06        $ 7.08M       $ 7.08M  
23532   9986090.82  6.830290e+06        $ 6.83M       $ 6.83M  
26518  12486197.52  6.653470e+06        $ 6.65M       $ 6.65M  

[5 rows x 26 columns]